### PR TITLE
Record all functions that use the newly built image

### PR DESCRIPTION
**Pull request by jasonmk**
#### Which issue(s) does this change fix?
#2466 

#### Why is this change necessary?

Currently, if multiple functions have the same Docker build metadata, only the first one will get an ImageUri tag in the updated template.yaml.  As a result, package will fail.

#### How does it address the issue?

This causes the resulting DockerTag to get recorded as "built" for all of the functions it applies to instead of just the first one.

#### What side effects does this change have?

During package, it will try to upload the image once for each function that uses it.  Because of how Docker works though, only a single image actually ends up in ECR.  It's more a visual annoyance and a small performance hit during packaging.

#### Checklist

- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [x] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by moelasmar — CHANGES_REQUESTED** [review]
Can you add unit/integration test cases that cover this change.

**Comment by jasonmk** [comment]
I think I've written an integration test, but I can't get the tests to run.  However, they won't run on the develop branch for me either.  Do I need to do something special to get them working?

`=================================================================================================================================== ERRORS ===================================================================================================================================
_____________________________________________________________________________________________ ERROR collecting tests/integration/local/invoke/runtimes/test_with_runtime_zips.py _____________________________________________________________________________________________
../../.pyenv/versions/3.7.2/lib/python3.7/unittest/loader.py:235: in getTestCaseNames
    testFnNames = list(filter(shouldIncludeMethod, dir(testCaseClass)))
../../.pyenv/versions/3.7.2/lib/python3.7/unittest/loader.py:232: in shouldIncludeMethod
    fullName = '%s.%s' % (testCaseClass.__module__, testFunc.__qualname__)
E   AttributeError: 'MarkDecorator' object has no attribute '__qualname__'
========================================================================================================================== short test summary info ===========================================================================================================================
ERROR tests/integration/local/invoke/runtimes/test_with_runtime_zips.py::TestWithDifferentLambdaRuntimeZips - AttributeError: 'MarkDecorator' object has no attribute '__qualname__'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!`

**Comment by jasonmk** [comment]
Just checking in. Is there some magic formula I'm missing in order to run the integration tests?

**Review by qingchm — CHANGES_REQUESTED** [review]
Happy to give a review! First of all can we look at any related unit tests to update or have a specific unit test designed to test this code piece? Thanks!

**Comment by CoshUS** [comment]
> AttributeError: 'MarkDecorator' object has no attribute '**qualname**'

For the integration test issue, it seems to be related to https://github.com/aws/aws-sam-cli/issues/1624.
Can you try creating a new python virtual environment?

File: samcli/lib/build/build_strategy.py [hunk 1]
```diff
@@ -111,7 +111,8 @@ def build_single_function_definition(self, build_definition):
                              single_build_dir,
                              build_definition.metadata
         )
-        function_build_results[single_function_name] = result
+        for function in build_definition.functions:
+            function_build_results[function.name] = result
```
Comment (hawflau) [inline samcli/lib/build/build_strategy.py:115]: Looking at the merge conflict, it seems this change might break the change made in this [PR](https://github.com/aws/aws-sam-cli/pull/2623).

**Review by hawflau — CHANGES_REQUESTED** [review]
Apologies for us not being able to review the PR for over a year! And now the PR requires more work than it did...

@jasonmk would you be able to address the merge conflicts? If not, we will take over the PR and prioritize it internally.

**Comment by hawflau** [comment]
Switching to draft due to inactivity. Assuming the author is not available to address the change requests, we will prioritize it internally and pick up what's left.

**Comment by jfuss** [comment]
Closing as inactive. 

Happy to review an updated PR from the community, if someone wants to revive the PR

**Pull request closed without merging**